xbmc/utils/UnicodeConverter.cpp

```diff
@@ -0,0 +1,992 @@
+ */
+
+constexpr int CUnicodeConverter::m_Utf8CharMaxSize = 4;
+
+// Define Converters.
+//
+// Limited testing suggests that creating a converter is fast,
+// but perhaps the first access is expensive. It also appears that copies of
+// converters are not that large. Therefore create a thread_local array
+// of all of the Unicode converters that we may use. Making them thread_local
+// means that we don't have to worry about locking, with the resulting blocking
+// of code waiting for a converter that another thread is using.
+
+/*!
+ * \brief retrieves the specified iconv converter from our thread_local collection
+ *
+ * \param converterType specifies the characteristics of the converter
+ *
+ * \return an iconv converter
+ */
+thread_local std::vector<CUnicodeConverterType> converters;
+
+CUnicodeConverterType& CUnicodeConverter::CInnerUCodeConverter::GetConverter(
+    StdConversionType converterType)
+{
+  if (converters.size() == 0)
+  {
+    converters.emplace_back(CUnicodeConverterType{UTF8_SOURCE, UTF32_CHARSET});
+    converters.emplace_back(
+        CUnicodeConverterType{UTF32_CHARSET, "UTF-8", CUnicodeConverter::m_Utf8CharMaxSize});
+    converters.emplace_back(CUnicodeConverterType{UTF32_CHARSET, WCHAR_CHARSET});
+    converters.emplace_back(CUnicodeConverterType{WCHAR_CHARSET, UTF32_CHARSET});
+    converters.emplace_back(
+        CUnicodeConverterType{WCHAR_CHARSET, "UTF-8", CUnicodeConverter::m_Utf8CharMaxSize});
+    converters.emplace_back(CUnicodeConverterType{UTF8_SOURCE, WCHAR_CHARSET});
+  }
+  CUnicodeConverterType& myConverter = converters[converterType];
+  return myConverter;
+}
+
+template<typename ST_IN_T_ARG,
+         typename ST_OUT_T_ARG,
+         typename ST_IN_T, //= typename enable_if_basic_string_view_t<ST_IN_T_ARG>,
+         typename ST_OUT_T, // = typename enable_if_basic_string_t<ST_OUT_T_ARG>,
+         typename IN_CHAR_T, // = typename ST_IN_T_ARG::value_type,
+         typename OUT_CHAR_T> //= typename ST_OUT_T_ARG::value_type>
+
+ST_OUT_T CUnicodeConverter::CInnerUCodeConverter::stdConvert(
+    const enum StdConversionType convertType,
+    const ST_IN_T input,
+    const bool failOnInvalidChar, /* = false */
+    const bool substituteBadCodepoints /* = true */)
+{
+  if (input.empty())
+    return ST_OUT_T{};
+
+  if (convertType < 0 || convertType >= NumberOfStdConversionTypes)
+    return ST_OUT_T{}; // Empty string on error
+
+  CUnicodeConverterType& converter =
+      CUnicodeConverter::CInnerUCodeConverter::GetConverter(convertType);
+
+  return convert<ST_IN_T, ST_OUT_T>(converter.PrepareConverter(),
+                                    converter.GetTargetSingleCharMaxLen(), input, failOnInvalidChar,
+                                    substituteBadCodepoints);
+}
+
+struct CharPtrPtrAdapter
+{
+  /* Ensures the char buffers fed to iconv are cast to the correct
+   * type for the platform.
+   *
+   * iconv may declare inbuf to be char** rather than const char** depending
```

Comment:
why not just always cast to char**?

Reply:
This came from CharSetConverter and looked nasty, so I left it alone.
But after playing with it  a while I fail to see how we can avoid  doing essentially the same thing. If you always cast to char** then calls to any iconv implementation that  requires const char** will break, such is the one in tools/target/depends:
`size_t iconv (iconv_t cd,
const char* * inbuf, size_t * inbytesleft,
char* * outbuf, size_t * outbytesleft);`

Or am I missing something here? The iconv used on Linux does not declare inbut with a const:
`extern size_t iconv (iconv_t __cd, char **__restrict __inbuf,
		     size_t *__restrict __inbytesleft,
		     char **__restrict __outbuf,
		     size_t *__restrict __outbytesleft);`